File: 02/solution.py
RESULTS = {
    (1, 1): 3,
    (2, 1): 1,
    (3, 1): 2,
    (1, 3): 2,
    (2, 3): 3,
    (3, 3): 1,
}


with open('input.txt') as fp:
    data = [
        (ord(a) - 64, ord(b) - 87) for a, b in map(
            lambda x: tuple(x.split(' ')),
            [x for x in fp.read().strip().split('\n')],
        )
    ]


def get_score(data):
    return sum([b + (3 if a == b else (6 if b == (a % 3) + 1 else 0)) for a, b in data])


# Part I
print(get_score(data))

## Part II
print(get_score(map(lambda x: (x[0], x[0]) if x[1] == 2 else (x[0], RESULTS[x]), data)))

File: 02/Solution.hs
import Data.Char

updateResult (x, 2) = (x, x)
updateResult (1, 1) = (1, 3)
updateResult (2, 1) = (2, 1)
updateResult (3, 1) = (3, 2)
updateResult (1, 3) = (1, 2)
updateResult (2, 3) = (2, 3)
updateResult (3, 3) = (3, 1)

score d = sum (map scorePerMatch d)
    where scorePerMatch (a, b) | a == b = 3 + b
                               | b == (a `mod` 3) + 1 = 6 + b
                               | otherwise = b

parse raw = map (\x -> (ord (x !! 0) - 64, ord (x !! 2) - 87)) (lines raw)

solve = ((\x -> (part1 x, part2 x)) . parse) <$> readFile "input.txt"
    where part1 = score
          part2 = score . (map updateResult)
